internal/platform/select_windows.go

```diff
@@ -23,9 +39,97 @@ func syscall_select(n int, r, w, e *FdSet, timeout *time.Duration) (int, error)
 		return 0, nil
 	}
 	if r.IsSet(wasiFdStdin) {
+		fileType, err := syscall.GetFileType(syscall.Stdin)
+		if err != nil {
+			return 0, err
+		}
+		if fileType&syscall.FILE_TYPE_CHAR != 0 {
+			res, err := pollNamedPipe(context.TODO(), syscall.Stdin, timeout)
+			if err != nil {
+				return -1, err
+			}
+			if !res {
+				r.Zero()
+				return 0, nil
+			}
+		}
 		r.Zero()
 		r.Set(wasiFdStdin)
 		return 1, nil
 	}
 	return -1, syscall.ENOSYS
 }
+
+// pollNamedPipe polls the given named pipe handle for the given duration.
+//
+// The implementation actually polls every 100 milliseconds until it reaches the given duration.
+// The duration may be nil, in which case it will wait undefinely. The given ctx is
+// used to allow for cancellation. Currently used only in tests.
+func pollNamedPipe(ctx context.Context, pipeHandle syscall.Handle, duration *time.Duration) (bool, error) {
+	// Short circuit when the duration is zero.
+	if duration != nil && *duration == time.Duration(0) {
+		return peekNamedPipe(pipeHandle)
+	}
+	// Ticker that emits at every pollInterval.
+	tick := time.NewTicker(pollInterval)
+	defer tick.Stop()
+	// If the duration is nil, then poll forever.
+	if duration == nil {
+		for {
+			select {
+			case <-ctx.Done():
+				return false, nil
+			case <-tick.C:
+				res, err := peekNamedPipe(pipeHandle)
+				if err != nil {
+					return false, err
+				}
+				if res {
+					return res, nil
+				}
+			}
+		}
+	} else {
+		// Otherwise, leave after the given duration, and check every pollInterval.
+		after := time.NewTimer(*duration)
```

Comment:
I wonder if you can make this an empty timer and avoid the if/else copy/pasta? So when there is no duration a timer that never stops?

Not sure if this would work.
```
var neverEnding =  &time.Ticker{C: make(chan time.Time)}
```

Reply:
Receiving from the `nil` channel blocks forever (and thus never satisfies `select`). Your `var` can be the channel, and you initialize it inside an `if` otherwise it's `nil`.